ethcore/src/state.rs

```diff
@@ -217,6 +232,14 @@ impl State {
 
 	/// Mutate storage of account `address` so that it is `value` for `key`.
 	pub fn storage_at(&self, address: &Address, key: &H256) -> H256 {
+		// check local cache first
+		if let Some(value) = self.cache.borrow().get(address).and_then(|a|
+			match *a {
+				AccountEntry::Cached(ref account) => account.modified_storage_at(key),
```

Comment:
Storage key search and update now works like this:
1. If there's an entry for the account in the local cache check for the key and return it if found.
2. If there's an entry for the account in the global cache check for the key or load it into that account.
3. If account is missing in the global cache load it into the local cache and cache the key there.